core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -847,6 +847,301 @@ KOKKOS_INLINE_FUNCTION_DELETED Impl::TeamThreadRangeBoundariesStruct<
+  using MDTeamNextRanks =
+      MDTeamRangeRanks<Rank, ParThreadRank, ParVectorRank, nextRank>;
+  using MDTeamNextMode = typename MDTeamNextRanks::RangeMode;
+
+  for (int i = 0; i != policy.boundaries[CurrentRank]; ++i) {
+    if constexpr (Rank::outer_direction == Iterate::Right) {
+      nested_loop(MDTeamNextMode(), MDTeamNextRanks(), policy, lambda, val,
+                  args..., i);
+    } else {
+      nested_loop(MDTeamNextMode(), MDTeamNextRanks(), policy, lambda, val, i,
+                  args...);
+    }
+  }
+}
+
+template <MDTeamRangeParThread ParThread, MDTeamRangeParVector ParVector,
+          typename Rank, int ParThreadRank, int ParVectorRank, int CurrentRank,
+          typename Policy, typename Lambda, typename ReducerValueType,
+          typename... Args>
+KOKKOS_INLINE_FUNCTION void nested_loop(
+    MDTeamRangeMode<MDTeamRangeEndRank::NotEndRank, ParThread, ParVector> const&
+        mode,
+    MDTeamRangeRanks<Rank, ParThreadRank, ParVectorRank, CurrentRank>,
+    Policy const& policy, Lambda const& lambda, ReducerValueType& val,
+    Args... args) {
+  constexpr int nextRank =
+      CurrentRank + (Rank::outer_direction == Iterate::Right ? 1 : -1);
+  using MDTeamNextRanks =
+      MDTeamRangeRanks<Rank, ParThreadRank, ParVectorRank, nextRank>;
+  using MDTeamNextMode = typename MDTeamNextRanks::RangeMode;
+
+  // This recursively processes ranks from [0..TotalRank-1]
+  // args... is passed by value because it should always be ints
+  parallel_for(nested_policy(mode, policy.team, policy.boundaries[CurrentRank]),
+               [&](int const& i) {
+                 if constexpr (Rank::outer_direction == Iterate::Right) {
+                   nested_loop(MDTeamNextMode(), MDTeamNextRanks(), policy,
+                               lambda, val, args..., i);
+                 } else {
+                   nested_loop(MDTeamNextMode(), MDTeamNextRanks(), policy,
+                               lambda, val, i, args...);
+                 }
+               });
+}
+
+template <typename Rank, MDTeamRangeThreadAndVector threadAndVector,
+          typename MDTeamPolicy, typename Lambda, typename ReductionValueType,
+          typename Status>
+KOKKOS_INLINE_FUNCTION void md_parallel_impl(MDTeamPolicy const& policy,
+                                             Lambda const& lambda,
+                                             Status const& rangeStatus,
+                                             ReductionValueType& val) {
+  constexpr auto TotalRank = MDTeamPolicy::total_rank;
+  constexpr auto Iter      = MDTeamPolicy::iter;
+  using TeamHandle         = typename MDTeamPolicy::team_handle_type;
+
+  static_assert(MDTeamPolicy::total_rank >= 2 && MDTeamPolicy::total_rank <= 8);
+
+  using par_rank_t =
+      Impl::ParallelRank<Rank, typename TeamHandle::execution_space,
+                         threadAndVector>;
+
+  constexpr int beginRank = (Iter == Iterate::Right) ? 0 : (TotalRank - 1);
```

Comment:
Changed all cases to PascalCases if types or template parameters and to snake_cases if variables in [7361243](https://github.com/kokkos/kokkos/pull/5238/commits/7361243c9f3e76c1328b1c1d27df3f25f0ec410a).

Do we have a common naming style that we try to follow? Since Kokkos is already strictly enforcing a formatting style,  attempting to have a similar naming style could also be nice for consistency.